### PR TITLE
flow control: side stream flow control part 1

### PR DESCRIPTION
**Pull request by tyxia**
Description:
This PR implements flow control (part 1) : side stream applies back pressure to downstream. 
It hooks up the watermark callbacks and registration in envoy-gRPC.  When it is above buffer watermark, the high watermark callback will fire; watermark event will be populated and eventually inform the downstream (data source) stop sending the data.

Tests:
Envoy-gRPC test: Verifies that expected high/low watermark callbacks are triggered when large request is sent.
ext_proc test: Verifies that expected flow control stats are triggered (by watermark callbacks and events) when large request is sent.

### CODE REVIEW

**Comment by tyxia** [comment]
/assign @yanavlasov @htuch @alyssawilk


PTAL, Thanks!

File: envoy/grpc/async_client.h
Comment (alyssawilk) [inline envoy/grpc/async_client.h:78]: ENVOY_BUG?

File: source/common/grpc/google_async_client_impl.h
Comment (alyssawilk) [inline source/common/grpc/google_async_client_impl.h:241]: optonal, but do we not care about this internally as well though?

File: source/extensions/filters/http/ext_proc/client.h
Comment (alyssawilk) [inline source/extensions/filters/http/ext_proc/client.h:46]: can we get away with watermark callbacks?

File: source/extensions/filters/http/ext_proc/client_impl.cc
Comment (alyssawilk) [inline source/extensions/filters/http/ext_proc/client_impl.cc:35]: optional but I tend to runtime guard all things watermarks since they're complicated 
this one seems pretty straight-forward so your call

**Review by alyssawilk — COMMENTED** [review]
Very nice PR.  What do you think of adding the new hooks into the envoy flow control docs?
/wait

**Review by tyxia — COMMENTED** [review]
@alyssawilk  Thanks for review!

> Very nice PR. What do you think of adding the new hooks into the envoy flow control docs?

This PR is first part of sidestream control (sidestream push back downstream) and there will be at least 3 PRs afterwards (sidestream pushback upstream; downstream push back sidestream; upstream push back sidestream). 

The watermark API will be updated/changed in next PRs, as [existing API ](https://github.com/envoyproxy/envoy/blob/f0201e54683875efeecf09df7328ad374be52d2c/envoy/http/async_client.h#L194)  which register `DecoderFilterWatermarkCallbacks` can only apply back pressure to downstream through `onEncoderFilterAboveWriteBufferHighWatermark` but this work will need 4-directional push-back

Thus, my plan is to add them to flow control doc when whole sidestream flow control works is done or at least the new watermark API in future PRs are approved. WDYT about this plan?

File: source/extensions/filters/http/ext_proc/client_impl.cc
Comment (alyssawilk) [reply source/extensions/filters/http/ext_proc/client_impl.cc]: one flag SGTM but let's add it now so the new code is protected as we land it?  I think we'll need to latch a bool to make sure we consistently register and clear callbacks

File: envoy/grpc/async_client.h
Comment (alyssawilk) [inline envoy/grpc/async_client.h:78]: Ah, I didn't realize the assert was already in there.  honestly it's a good one to have as a bug (the code predated the macro) so thanks for the change :-)

**Review by alyssawilk — COMMENTED** [review]
lgtm modulo guard and release note

File: source/extensions/filters/http/ext_proc/client_impl.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/client_impl.cc]: Sure SG! Done.

File: source/extensions/filters/http/ext_proc/client_impl.cc
Comment (alyssawilk) [inline source/extensions/filters/http/ext_proc/client_impl.cc:37]: I think we need to latch this to make sure we don't end up wedged if we flip midway through stream lifetime
/wait

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/client_impl.cc]: make sense. Done.

PTAL, thanks!

**Comment by alyssawilk** [comment]
adding cross-company reviewer for at least the first PR in the series - I'll assume LGTM here is LGTM for the four-fold flow control integration unless otherwise specified.

File: envoy/grpc/async_client.h
Comment (wbpcode) [inline envoy/grpc/async_client.h:86]: I personally prefer to use the StreamOptions as the unique channel to propagate the context, callbacks, options between the main stream and side stream.

But considering that way may effects the basic implementation of grpc client and the http stream has used current way, so, I won't block this PR. )

**Review by wbpcode — APPROVED** [review]
Thanks for the contribution. LGTM.

**Pull request merged**
